src/supernova/mod.rs

```diff
@@ -176,92 +176,182 @@ where
+  /// realign them if necessary. If a digest is given, we assume that it
+  /// will correctly match the public params after realigning them, and
+  /// we do not check for its validity.
+  pub fn from_pp_vec(pp_vec: Vec<PublicParams<G1, G2>>, digest: OnceCell<G1::Scalar>) -> Self {
+    let num_pp = pp_vec.len();
+    let primary_len = &pp_vec[0].ck_primary.as_ref().map_or(0, Len::length);
+    let secondary_len = &pp_vec[0].ck_secondary.as_ref().map_or(0, Len::length);
+    let num_complete = pp_vec
+      .iter()
+      .filter(|pp| {
+        pp.ck_primary.is_some()
+          && pp.ck_secondary.is_some()
+          && &pp.ck_primary.as_ref().unwrap().length() == primary_len
+          && &pp.ck_secondary.as_ref().unwrap().length() == secondary_len
+      })
+      .count();
+
+    let stage = if num_complete == num_pp {
+      Stage::Complete
+    } else {
+      Stage::Incomplete
+    };
+
+    let mut running_claim_params = RunningClaimParams {
+      pp_vec,
+      digest,
+      stage,
+      _p: PhantomData,
+    };
+
+    running_claim_params.complete();
+    running_claim_params
+  }
 
-    for claim in &mut claims {
-      claim.set_commitment_key(ck_primary.clone(), ck_secondary.clone());
+  /// Compute primary and secondary commitment keys sized to handle the largest of the circuits in the provided
+  /// `PublicParams`.
+  fn compute_commitment_keys(&self) -> (CommitmentKey<G1>, CommitmentKey<G2>) {
+    macro_rules! max_size {
+      ($shape_getter:ident, $ck_getter:ident) => {
+        self
+          .pp_vec
+          .iter()
+          .map(|pp| {
+            let ck_len = if let Some(ck) = &pp.$ck_getter {
+              ck.length()
+            } else {
+              0
+            };
+            max(ck_len, commitment_key_size(&pp.$shape_getter, None))
+          })
+          .max()
+          .unwrap()
+      };
     }
-    RunningClaims {
-      claims,
-      digest: OnceCell::new(),
+
+    let size_primary = max_size!(r1cs_shape_primary, ck_primary);
+    let size_secondary = max_size!(r1cs_shape_secondary, ck_secondary);
+
+    let ck_primary = G1::CE::setup(b"ck", size_primary);
+    let ck_secondary = G2::CE::setup(b"ck", size_secondary);
+
+    (ck_primary, ck_secondary)
+  }
+
+  /// Completes the set of public params
+  pub fn complete(&mut self) {
+    if matches!(self.stage, Stage::Complete) {
+      return;
     }
+
+    self.set_commitment_keys();
+    self.digest();
   }
 
-  /// Return the `RunningClaims`' digest.
+  /// set primary/secondary commitment key for all the params
+  fn set_commitment_keys(&mut self) {
+    if matches!(self.stage, Stage::Complete) {
+      return;
+    }
+    let (ck_primary, ck_secondary) = self.compute_commitment_keys();
+    for pp in self.pp_vec.iter_mut() {
+      pp.ck_primary = Some(ck_primary.clone()); // TODO: the keys should be shared across all the params
+      pp.ck_secondary = Some(ck_secondary.clone()); // TODO: the keys should be shared across all the params
```

Comment:
Hint for why I think this deserves more thought:
- the CKs are ginormous,
- they are a universal parameter, i.e. they are not circuit-dependent. They depend on the size of the circuit (num of variables and constraints, and matrix size, see #5), but can be shared between any circuits that are under the max size that was passed at their creation. See #5.

Reply:
I attempted this by moving all the shared stuff into the topmost struct. It's a bit confusing since I've renamed everything, so please take a look at the code.